### PR TITLE
Hierarchical multi-label classification

### LINKED ISSUE #121: hierarchical multilabel classification head
Hierarchical Multilabel Classification (HMC) provides improvement over naive classification in a lot of cases.

This requires to Implement the HMC head and loss function like in [C-HMCNN](https://github.com/EGiunchiglia/C-HMCNN)

This includes
- [x] consider the best hierarchical data format for ~~outcome~~ dataset (like [`data.tree`](https://cran.r-project.org/web/packages/data.tree/vignettes/data.tree.html#introduction) ~~or [`tidygraph`](https://tidygraph.data-imaginist.com/)~~ )
- [x] choose an example dataset
- [x] create a `tabnet_fit.Node()` method
- [x] allow for `valid_split =` in tabnet_fit.Node()
- [x] fix a `tabnet_pretrain.Node()` void method
- [x] create a `predict.tabnet()` method for Node
- [x] create a `tabnet_explain()` method for Node()
- [x] create a predict method for data = Node
- [x] implement the R matrix computing from data.tree
- [x] implement the MCM loss
- [ ] create an inspiring vignette

### LINKED ISSUE #125: multi-outcome recipe fails with wierd message @tab-network.R#707 `index out of range in self`
# current situation

One multi-outcome recipe fails even if the multi-outcome recipe attrition test is passing :
https://github.com/mlverse/tabnet/blob/c60fbfd4239f848a0096154361ba13f39983c1fd/tests/testthat/test-hardhat_multi-outcome.R#L95-L115

# expected situation

No such situation

# ReprEx
``` r
library(tabnet)
library(recipes)
#> Le chargement a nécessité le package : dplyr
#> 
#> Attachement du package : 'dplyr'
#> Les objets suivants sont masqués depuis 'package:stats':
#> 
#>     filter, lag
#> Les objets suivants sont masqués depuis 'package:base':
#> 
#>     intersect, setdiff, setequal, union
#> 
#> Attachement du package : 'recipes'
#> L'objet suivant est masqué depuis 'package:stats':
#> 
#>     step
starwars_rec <-  recipe(species + homeworld ~ ., data = starwars %>% select(-where(is.list))) %>%
    step_impute_mode(all_outcomes()) %>%
    step_string2factor(all_string_predictors(), all_outcomes())

fit_multi <- tabnet_fit(starwars_rec, starwars, cat_emb_dim = 2)
#> Error in (function (weight, indices, padding_idx, scale_grad_by_freq, : index out of range in self
#> Exception raised from operator() at ../aten/src/ATen/native/TensorAdvancedIndexing.cpp:1194 (most recent call first):
#> frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x6b (0x7f338bca6d4b in /home/creg/R/x86_64-pc-linux-gnu-library/4.3/torch/lib/libc10.so)
#> frame #1: <unknown function> + 0x197e491 (0x7f337257e491 in /home/creg/R/x86_64-pc-linux-gnu-library/4.3/torch/lib/libtorch_cpu.so)
#> frame #2: at::native::index_select_out_cpu_(at::Tensor const&, long, at::Tensor const&, at::Tensor&) + 0x1a00 (0x7f33725a6d30 in /home/creg/R/x86_64-pc-linux-gnu-library/4.3/torch/lib/libtorch_cpu.so)
#> frame #3: at::native::index_select_cpu_(at::Tensor const&, long, at::Tensor const&) + 0x70 (0x7f33725a7d40 in /home/creg/R/x86_64-pc-linux-gnu-library/4.3/torch/lib/libtorch_cpu.so)
#> frame #4: <unknown function> + 0x24535f6 (0x7f33730535f6 in /home/creg/R/x86_64-pc-linux-gnu-library/4.3/torch/lib/libtorch_cpu.so)
#> frame #5: at::_ops::index_select::call(at::Tensor const&, long, at::Tensor const&) + 0x172 (0x7f3372aa68a2 in /home/creg/R/x86_64-pc-linux-gnu-library/4.3/torch/lib/libtorch_cpu.so)
#> frame #6: at::native::embedding(at::Tensor const&, at::Tensor const&, long, bool, bool) + 0x6c6 (0x7f3372302836 in /home/creg/R/x86_64-pc-linux-gnu-library/4.3/torch/lib/libtorch_cpu.so)
#> frame #7: <unknown function> + 0x25f8ad2 (0x7f33731f8ad2 in /home/creg/R/x86_64-pc-linux-gnu-library/4.3/torch/lib/libtorch_cpu.so)
#> frame #8: at::_ops::embedding::redispatch(c10::DispatchKeySet, at::Tensor const&, at::Tensor const&, long, bool, bool) + 0xb1 (0x7f3372f5d291 in /home/creg/R/x86_64-pc-linux-gnu-library/4.3/torch/lib/libtorch_cpu.so)
#> frame #9: <unknown function> + 0x3b03e6a (0x7f3374703e6a in /home/creg/R/x86_64-pc-linux-gnu-library/4.3/torch/lib/libtorch_cpu.so)
#> frame #10: <unknown function> + 0x3b04435 (0x7f3374704435 in /home/creg/R/x86_64-pc-linux-gnu-library/4.3/torch/lib/libtorch_cpu.so)
#> frame #11: at::_ops::embedding::call(at::Tensor const&, at::Tensor const&, long, bool, bool) + 0x1a5 (0x7f3372fa6095 in /home/creg/R/x86_64-pc-linux-gnu-library/4.3/torch/lib/libtorch_cpu.so)
#> frame #12: at::embedding(at::Tensor const&, at::Tensor const&, long, bool, bool) + 0x5c (0x7f338aae6e04 in /home/creg/R/x86_64-pc-linux-gnu-library/4.3/torch/lib/liblantern.so)
#> frame #13: _lantern_embedding_tensor_tensor_intt_bool_bool + 0x17b (0x7f338a629110 in /home/creg/R/x86_64-pc-linux-gnu-library/4.3/torch/lib/liblantern.so)
#> frame #14: cpp_torch_namespace_embedding_weight_Tensor_indices_Tensor(XPtrTorchTensor, XPtrTorchIndexTensor, XPtrTorchint64_t, XPtrTorchbool, XPtrTorchbool) + 0x41 (0x7f338b8a0bd1 in /home/creg/R/x86_64-pc-linux-gnu-library/4.3/torch/libs/torchpkg.so)
#> frame #15: _torch_cpp_torch_namespace_embedding_weight_Tensor_indices_Tensor + 0xed (0x7f338b65342d in /home/creg/R/x86_64-pc-linux-gnu-library/4.3/torch/libs/torchpkg.so)
#> frame #16: <unknown function> + 0xfab5e (0x7f33966fab5e in /usr/lib/R/lib/libR.so)
#> frame #17: <unknown function> + 0x13dde9 (0x7f339673dde9 in /usr/lib/R/lib/libR.so)
#> frame #18: Rf_eval + 0x88 (0x7f3396751c78 in /usr/lib/R/lib/libR.so)
#> frame #19: <unknown function> + 0x15403e (0x7f339675403e in /usr/lib/R/lib/libR.so)
#> frame #20: Rf_applyClosure + 0x1a2 (0x7f3396754eb2 in /usr/lib/R/lib/libR.so)
#> frame #21: Rf_eval + 0x2af (0x7f3396751e9f in /usr/lib/R/lib/libR.so)
#> frame #22: <unknown function> + 0xc4173 (0x7f33966c4173 in /usr/lib/R/lib/libR.so)
#> frame #23: <unknown function> + 0x1379bb (0x7f33967379bb in /usr/lib/R/lib/libR.so)
#> frame #24: Rf_eval + 0x88 (0x7f3396751c78 in /usr/lib/R/lib/libR.so)
#> frame #25: <unknown function> + 0x15403e (0x7f339675403e in /usr/lib/R/lib/libR.so)
#> frame #26: Rf_applyClosure + 0x1a2 (0x7f3396754eb2 in /usr/lib/R/lib/libR.so)
#> frame #27: <unknown function> + 0x14001e (0x7f339674001e in /usr/lib/R/lib/libR.so)
#> frame #28: Rf_eval + 0x88 (0x7f3396751c78 in /usr/lib/R/lib/libR.so)
#> frame #29: <unknown function> + 0x15403e (0x7f339675403e in /usr/lib/R/lib/libR.so)
#> frame #30: Rf_applyClosure + 0x1a2 (0x7f3396754eb2 in /usr/lib/R/lib/libR.so)
#> frame #31: <unknown function> + 0x14001e (0x7f339674001e in /usr/lib/R/lib/libR.so)
#> frame #32: Rf_eval + 0x88 (0x7f3396751c78 in /usr/lib/R/lib/libR.so)
#> frame #33: <unknown function> + 0x15403e (0x7f339675403e in /usr/lib/R/lib/libR.so)
#> frame #34: Rf_applyClosure + 0x1a2 (0x7f3396754eb2 in /usr/lib/R/lib/libR.so)
#> frame #35: <unknown function> + 0x14001e (0x7f339674001e in /usr/lib/R/lib/libR.so)
#> frame #36: Rf_eval + 0x88 (0x7f3396751c78 in /usr/lib/R/lib/libR.so)
#> frame #37: <unknown function> + 0x15403e (0x7f339675403e in /usr/lib/R/lib/libR.so)
#> frame #38: Rf_applyClosure + 0x1a2 (0x7f3396754eb2 in /usr/lib/R/lib/libR.so)
#> frame #39: <unknown function> + 0x14001e (0x7f339674001e in /usr/lib/R/lib/libR.so)
#> frame #40: Rf_eval + 0x88 (0x7f3396751c78 in /usr/lib/R/lib/libR.so)
#> frame #41: <unknown function> + 0x15403e (0x7f339675403e in /usr/lib/R/lib/libR.so)
#> frame #42: Rf_applyClosure + 0x1a2 (0x7f3396754eb2 in /usr/lib/R/lib/libR.so)
#> frame #43: Rf_eval + 0x2af (0x7f3396751e9f in /usr/lib/R/lib/libR.so)
#> frame #44: <unknown function> + 0x155c40 (0x7f3396755c40 in /usr/lib/R/lib/libR.so)
#> frame #45: Rf_eval + 0x58b (0x7f339675217b in /usr/lib/R/lib/libR.so)
#> frame #46: <unknown function> + 0x15403e (0x7f339675403e in /usr/lib/R/lib/libR.so)
#> frame #47: Rf_applyClosure + 0x1a2 (0x7f3396754eb2 in /usr/lib/R/lib/libR.so)
#> frame #48: Rf_eval + 0x2af (0x7f3396751e9f in /usr/lib/R/lib/libR.so)
#> frame #49: <unknown function> + 0x1561d3 (0x7f33967561d3 in /usr/lib/R/lib/libR.so)
#> frame #50: Rf_eval + 0x58b (0x7f339675217b in /usr/lib/R/lib/libR.so)
#> frame #51: <unknown function> + 0x155c40 (0x7f3396755c40 in /usr/lib/R/lib/libR.so)
#> frame #52: Rf_eval + 0x58b (0x7f339675217b in /usr/lib/R/lib/libR.so)
#> frame #53: Rf_eval + 0x58b (0x7f339675217b in /usr/lib/R/lib/libR.so)
#> frame #54: <unknown function> + 0x155c40 (0x7f3396755c40 in /usr/lib/R/lib/libR.so)
#> frame #55: Rf_eval + 0x58b (0x7f339675217b in /usr/lib/R/lib/libR.so)
#> frame #56: <unknown function> + 0x15a4e9 (0x7f339675a4e9 in /usr/lib/R/lib/libR.so)
#> frame #57: Rf_eval + 0x58b (0x7f339675217b in /usr/lib/R/lib/libR.so)
#> frame #58: <unknown function> + 0x155c40 (0x7f3396755c40 in /usr/lib/R/lib/libR.so)
#> frame #59: Rf_eval + 0x58b (0x7f339675217b in /usr/lib/R/lib/libR.so)
#> frame #60: <unknown function> + 0x15403e (0x7f339675403e in /usr/lib/R/lib/libR.so)
#> frame #61: Rf_applyClosure + 0x1a2 (0x7f3396754eb2 in /usr/lib/R/lib/libR.so)
#> frame #62: Rf_eval + 0x2af (0x7f3396751e9f in /usr/lib/R/lib/libR.so)
#> frame #63: <unknown function> + 0x1571e2 (0x7f33967571e2 in /usr/lib/R/lib/libR.so)
```

<sup>Created on 2023-06-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by cregouby**
This P.R. adds 
- hierarchical inference based on `Node` dataset encoding via implementation of [Coherent Hierarchical Multi-label Classification Networks]. This has required an opinonated design change : a ancestor matrix (n by n booleans, n being the total nlevels() of the outcomes) to be piggy-back inside the tabnet `config` list.  
- renaming of "unsupervised learning" into "self-supervised learning" 

Fix #121 
Fix #125

### CODE REVIEW

File: vignettes/Hierarchical_classification.Rmd [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+vip::vip(starw_model)
+```
+
+## Model inference
+
+We can infer on the test-set
+```{r}
+starwars_hat <- bind_cols(
+    predict(starw_model, starwars_test_tree),
+    tabnet:::node_to_df(starwars_test_tree)$y
```
Comment (dfalbel) [inline vignettes/Hierarchical_classification.Rmd:172]: It looks like you are using an internal function here. Maybe export?

File: vignettes/Hierarchical_classification.Rmd [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+# erroneous Node construction
+starwars_tree <- starwars %>% 
+  mutate(pathString = paste("StarWars_characters", species, sex, `name`, sep = "/")) %>%
+  as.Node()
+print(starwars_tree, "name","height", "mass", "eye_color", limit = 8)
+
+```
+
+You may have notice that `name` and `height` have unexpected values according to the original data: `Human` is not part of the `name` in orginal dataset, and `height` values have been changed into the local height of the tree. This is due to some rules we will have to follow to create the Node from data frame.
```
Comment (dfalbel) [inline vignettes/Hierarchical_classification.Rmd:60]:
```suggestion
You may have noticed that `name` and `height` have unexpected values according to the original data: `Human` is not part of the `name` in orginal dataset, and `height` values have been changed into the local height of the tree. This is due to some rules we will have to follow to create the Node from data frame.
```

File: vignettes/Hierarchical_classification.Rmd [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+We can see in the Warnings that the dataset is a challenge as many new levels are found in a lot of predictors in the test set.  
+The model also here is very poor on the `level_2` ( `species` ) and on `level_3` ( ` sex` ) as this is definitively not a model-intended dataset. The reason is that the input dataset not collecting large samples of distinctive observation per leaf, but rather a very diverse but limited number of characters compatible with watching a movie saga.
+
+Despite the performance, we do have local feature importance on the complete dataset here :  
+```{r}
+starwars_explain <- tabnet_explain(starw_model, starwars_test_tree)
+autoplot(starwars_explain)
+autoplot(starwars_explain, type = "steps")
+```
+Hopefully your own hierarchical outcome will have a better success than the one here with `starwars` dataset. But in this journey, you have learn a lot in the data format constraints and solutions, and you now have a new performing solution in your toolbox.
```
Comment (dfalbel) [inline vignettes/Hierarchical_classification.Rmd:187]:
```suggestion
Hopefully your own hierarchical outcome will have a better success than the one here with `starwars` dataset. But in this journey, you have learned a lot in the data format constraints and solutions, and you now have a new performing solution in your toolbox.
```

File: R/model.R [hunk 4]
```diff
@@ -203,15 +204,32 @@ tabnet_config <- function(batch_size = 1024^2,
...
     loss_fn <- loss
   else if (loss %in% c("mse", "auto") && !dtype == torch::torch_long())
     loss_fn <- torch::nn_mse_loss()
-  else if (loss %in% c("bce", "cross_entropy", "auto") && dtype == torch::torch_long())
+  else if ((loss %in% c("bce", "cross_entropy", "auto") && dtype == torch::torch_long()) || !is.null(config$ancestor_tt))
+    # cross entropy loss is required
     loss_fn <- torch::nn_cross_entropy_loss()
   else
-    rlang::abort(paste0(loss," is not a valid loss for outcome of type ",dtype))
+    rlang::abort(glue::glue("{loss} is not a valid loss for outcome of type {dtype}"))
```
Comment (dfalbel) [inline R/model.R:232]: I think we need to add `glue` to the Imports in the DESCRIPTION.

**Review by dfalbel — COMMENTED** [review]
Looks great @cregouby ! Really nice :)

File: R/model.R (hunk 4 above)
Comment (cregouby) [reply R/model.R:232]: exported !
